Add RestrictFileSystems= property using LSM BPF

iaguis opened the pull request:
This PR adds the `RestrictFileSystems=` property. When used, processes
belonging to a service are only able to access the filesystems listed in the
property.

This is implemented by attaching a BPF program to the `file_open` BPF LSM hook.
The program is attached at boot time and stays there forever. Then, when a
service specifying the `RestrictFileSystems=` property is started, an entry is
added to a global hash of maps BPF map pinned to the BPF filesystem under
`/sys/fs/bpf/systemd/lsm_bpf_map`. The map stores a set of filesystem
magic numbers per cgroupID. When a process tries to open a file, the BPF
program is executed and checks the cgroup the process is running in: if an
entry is present in the global map it checks if the filesystem the process is
trying to access is present in the set, if not, it denies access to it.

`RestrictFileSystems=` is only supported on systems with the LSM BPF hook
enabled and using cgroup2 (unified or hybrid).

This PR makes use of the libbpf framework proposed on #17655. Same as that PR,
it requires clang and llvm at compile time, and the
[libbpf](https://github.com/libbpf/libbpf) shared library.

Thanks to the usage of libbpf, the program can use the CO-RE (Compile-Once
Run-Everywhere) technology so it doesn't require kernel headers at runtime to
access internal kernel structures.

gportay committed:
man: fix path reference to unit file

The unit files are located at path /usr/lib/systemd/system.

This fixes the path reference to the unit file by adding the missing
path component system.

yuwata committed:
network: set FRA_PROTOCOL to RTPROT_STATIC by default

yuwata committed:
network: do not serialize/deserialize routing policy rules

We already handle foreign routing policy rules correctly by the previous
commit. So, the serialization/deserialization of rules are not necessary
anymore.

yuwata committed:
network: drop unnecessary routing policy rules

networkd already drop foreign address, routes, and nexthops on startup,
except those created by kernel. However, previously, routing policy
rules were not. The logic of serialization/deserialization of rules only
works for rules created by previous invocation of networkd, and does not
work for one created by other tools like `ip rule`.

This makes networkd drop foreign routing policy rules except created by
kernel on startup. Also, remove rules created by networkd when the
corresponding links are dropped or networkd is stopping.

yuwata committed:
network: adjust protocol of rules sent from kernel when kernel does not support FRA_PROTOCOL

Otherwise, each configured rule is treated as foreign.

yuwata committed:
network: treat rule which has l3mdev flag as created by kernel

Commit:
bpf: add bpf_object helpers for source compiled bpf

Introduce minimalistic set of helpers for bpf programs compiled from
restricted C sources.

Introduce a basic type `struct BPFProgramV2` with 'fd'
and 'attach_type' fields to represent a loaded bpf prog:wqram.
The BPFProgram struct is not used since:
- v2 methods will use libbpf while v1 use raw syscalls
- the majority of its fields is not needed to support BPF program
compiled from sources
- lack of 'attach_type' field

Introduce bpf_object_{} helpers to load bpf programs into kernel, resize
and populate bpf maps, attach program to cgroup hooks.

libbpf dependency must be satisfied to compile the code.

Commit:
bpf: add build script for bpf programs

Add a build script to compile bpf source code. A program in restricted
C is compiled into a temporary object file.
Script generates a single C header file with defined
`const unsigned char[]` array (hexdump) representing object file: each
octet of the object file is an element of the array.
To generate C array script reads *.o file in byte mode in chunks of 16
bytes. Each byte is represented as hexadecimal number with 0x prefix
(hence hexdump), a chunk of 16 numbers forms a line of the target file,
e.g:
```
0x7f, 0x45, 0x4c, 0x46, 0x02, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00,
0x00, 0x00, 0x00, 0x00
```

Why hexdump?
- Ready in compile time. No need to distribute *.o files along with
systemd package, the bytecode of BPF program ships with `#include`.
- Transparent and stdout friendly.

In user space hexdump may be included as a regular C header and
read into `struct bpf_object [0] with `bpf_object__open_mem` helper [1].

If build with custom meson build rule, the target header will reside in
build/ directory (not in source tree), e.g the path for allow_bind:
`build/src/core/bpf/allow_bind/allow-bind-hexdump.h`

Summing up, script runs three phases:
* clang to generate LLVM *.bc from restricted C
* llc to compile *.o from *.bc input
* hexdump generation
These phases are logged to stderr for debug purposes.

To include BTF debug information, -g option is passed to clang.

Sample run for `src/core/bpf/allow_bind` program
```
./tools/build-bpf.py --llc_exec /usr/bin/llc --clang_exec /usr/bin/clang
src/core/bpf/allow_bind/allow-bind.c
src/core/bpf/allow_bind/allow-bind-hexdump.h --bpf_hexdump_buffer
allow_bind_hexdump_buffer
DEBUG:root:Generating LLVM bitcode *.bc:
DEBUG:root:/usr/bin/clang -Wno-compare-distinct-pointer-types -O2
-target bpf -emit-llvm -c -D__x86_64__ -I/usr/include/x86_64-linux-gnu/
-I/usr/local/include -I/usr/include -I.
src/core/bpf/allow_bind/allow-bind.c -o /tmp/tmpk8sgch4l.bc
DEBUG:root:Compiling BPF object file:
DEBUG:root:/usr/bin/llc -march=bpf -filetype=obj -o /tmp/tmpk6fxj7j_.o
/tmp/tmpk8sgch4l.bc
DEBUG:root:Generating hexdump for src/core/bpf/allow_bind/allow-bind.c
source from /tmp/tmpk6fxj7j_.o object file
```

[0] https://github.com/libbpf/libbpf/blob/master/src/libbpf.h#L61
[1] https://github.com/libbpf/libbpf/blob/master/src/libbpf.h#L103

Commit:
bpf: add HAVE_LIBBPF, BUILD_BPF config options and feature gate

* Add `build-bpf` feature gate with 'auto', 'true' and 'false' choices
* Add libbpf [0] dependency
* Search for clang and llc binaries the build environment.

For libbpf [0], make 0.2.0 [1] the minimum required version.
If libbpf is satisfied, set HAVE_LIBBPF config option to 1.

If `build-bpf` feature gate is set to 'auto', whether feature is enabled
or disabled is defined by presence of all of libbpf, clang and llvm in build
environment. With 'auto' all dependencies are optional.
If the gate is set to `true`, make all of the libbpf, clang and llvm
dependencies mandatory.
If it's set to `false`, set `BUILD_BPF` to false and make libbpf
dependency optional.

libbpf dep is dynamic followed by the common pattern in systemd.
find_program doesn't allow to set minimum version similary to
`dependency` option. The most recent BPF features include BTF which
require minimim v.10 LLVM, allow_bind program doesn't use BTF features
and builds with clang and llvm 9.0.

daandemeyer committed:
Enable more mkosi options

- HostonlyInitrd reduces the size of the initramfs which speeds up
QEMU boot times.
- Autologin and deleting the root password remove
the manual login step whenever starting the systemd-nspawn container
or QEMU vm.
- NetworkVeth passes --network-veth to the nspawn container when
using mkosi boot and emulates --network-veth with a tap device
and networkd's 80-vm-vt.network builtin network file when using
mkosi qemu. --network-veth makes it easier to test network and
resolved in mkosi's nspawn container or QEMU vm.

daandemeyer committed:
mkosi: Enable HostonlyInitrd option

Split out from #17919.

iaguis committed:
stat-util: add fs_type_from_string()

Returns the magic number for each filesystem.

iaguis committed:
bpf: add restrict_fs BPF program

It hooks into the file_open LSM hook and allows only when the filesystem
where the open will take place is present in a BPF map for a particular
cgroup.

The BPF map used is a hash of maps with the following structure:

    cgroupID -> (s_magic -> uint32)

The inner map is effectively a set.

When the cgroupID is present in the map, it checks the inner map for the
magic number of the filesystem associated with the file that's being
opened. If that magic number is present it allows the open to succeed,
otherwise it returns -EPERM.

If the cgroupID is not present in the map, it allows the open to
succeed.

The BPF program uses CO-RE (Compile-Once Run-Everywhere) to access
internal kernel structures without needing kernel headers present at
runtime.

iaguis committed:
bpf: add meson build rules for the restrict_fs program

It uses tools/build-bpf.py to compile the BPF program from the sources.

iaguis committed:
mkosi: add libbpf dependency

For distros that ship libbpf 0.2.0.

iaguis committed:
test: use restrict-fs BPF program to test bpf_object functions

If #17655 gets merged, there's no need to do this and we can use their
test.

This removes the bpf_object_get_programs() test because LSM programs
are not returned by libbpf.

iaguis committed:
cgroup-util: add cg_path_get_cgroupid()

It returns the cgroupID from a cgroup path.

iaguis committed:
core: add RestrictFileSystems= fragment parser

Services only have access to filesystems that are listed here.

Accepts a list of filesystem names.

iaguis committed:
core: add dbus RestrictFileSystems= properties

iaguis committed:
exit-status: fix mappings comment

It didn't reflect the current status.

iaguis committed:
exit-status: add EXIT_BPF

It will be used later.

iaguis committed:
man: add RestrictFileSystems= documentation

iaguis committed:
test: add libbpf dependency

libbpf is used in core code now, so we need to add it as dependency for
tests.

iaguis committed:
bpf-object: add set inner map fd and find prog by title functions

bpf_object_set_inner_map_fd is needed for hash of maps BPF maps and
bpf_object_find_program_by_title is needed because
bpf_object_get_programs() doesn't return LSM BPF programs, so we need to
get it by name.

iaguis committed:
meson: add libbpf deps for systemd and systemd-analyze

They link with libcore and libcore is not using libbpf.

iaguis committed:
core: add BPF LSM functions

This adds 4 functions to implement RestrictFileSystems=

* lsm_bpf_supported() checks if LSM BPF is supported. It checks that
  cgroupv2 is used, that BPF LSM is enabled, and tries to load the BPF
  LSM program which makes sure BTF and hash of maps are supported, and
  BPF LSM programs can be loaded.
* lsm_bpf_setup() loads and attaches the LSM BPF program.
* bpf_restrict_filesystems() populates the hash of maps BPF map with the
  cgroupID and the set of allowed filesystems.
* cleanup_lsm_bpf() removes a cgroupID entry from the hash of maps.

iaguis committed:
core: use LSM BPF functions to implement RestrictFileSystems=

It attaches the LSM BPF program when the system manager starts up.

It populates the hash of maps BPF map when services that have
RestrictFileSystems= set start.

It cleans up the hash of maps when the unit cgroups is pruned.

iaguis committed:
test: add test-bpf-lsm

alban commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,260 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to get systemd cgroup path");
+
+        r = cg_path_get_cgroupid(path, &cgroup_id);
```
On cgroup legacy, we cannot use the cgroupid in bpf. Either `cg_path_get_cgroupid()` should check that the path in on `CGROUP2_SUPER_MAGIC` or the callers of `cg_path_get_cgroupid()` should check it is only calling on a cgroup2.

The PR description should say what's the behaviour of `RestrictFileSystems=` when using cgroup legacy.

iaguis replied:
Before calling this function in execute.c, `lsm_bpf_supported()` is called which checks cgroup2 is supported so this will never be called in legacy mode. Do you think I should add a check in this function too?

I'll update the PR description and perhaps add this to the documentation :+1:

alban commented on src/core/bpf-lsm.h:
```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#define LSM_MAP_BPFFS_PATH "/sys/fs/bpf/systemd/lsm_bpf_map"
```
What's happening when systemd is started inside systemd-nspawn? Does systemd-nspawn prepare a new bpffs or does it bind mount the bpffs from the host? Does it work when several systemds uses the same `lsm_bpf_map` map?

iaguis replied:
Good point, I tested and it fails as follows:

```
Dec 17 13:47:04 image systemd[113]: test.service: Failed to restrict filesystems: Operation not supported
Dec 17 13:47:04 image systemd[113]: test.service: Failed at step BPF spawning /usr/bin/sleep: Operation not supported
Dec 17 13:47:04 image systemd[1]: test.service: Main process exited, code=exited, status=244/BPF
Dec 17 13:47:04 image systemd[1]: test.service: Failed with result 'exit-code'.
```

I see there's no `/sys/kernel/security/lsm` file inside the container.

Also, systemd-nspawn seems to mount a new bpffs.

alban commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,260 @@
+        if (bpf_map_update_elem(outer_map_fd, &cgroup_id, &inner_map_fd, BPF_ANY) != 0)
+                return log_error_errno(errno, "Error populating LSM BPF map: %m");
+
+
```
double empty line

alban commented on src/core/bpf-lsm.c:
```diff
@@ -0,0 +1,260 @@
+                        return r;
+
+                if (bpf_map_update_elem(inner_map_fd, &magic, &dummy_value, BPF_ANY) != 0)
+                        return log_error_errno(errno, "Failed to update BPF map: %m");
```
I guess you should cleanup the `bpf_map_delete_elem(outer_map_fd...)`  in case of failure (and that would automatically free the inner map).

iaguis replied:
Good point.

alban commented on src/core/bpf/restrict_fs/restrict-fs.c:
```diff
@@ -0,0 +1,58 @@
+                return 0;
+
+        /* Allow access only if magic_number present in inner map */
+        if (bpf_map_lookup_elem(magic_map, &magic_number) == 0)
```
`s/0/NULL`? (although it does not matter for the compiler)

iaguis replied:
OK

alban commented on src/shared/exit-status.c:
```diff
@@ -71,6 +71,7 @@ const ExitStatusMapping exit_status_mappings[256] = {
         [EXIT_CONFIGURATION_DIRECTORY] = { "CONFIGURATION_DIRECTORY", EXIT_STATUS_SYSTEMD },
         [EXIT_NUMA_POLICY] =             { "NUMA_POLICY",             EXIT_STATUS_SYSTEMD },
         [EXIT_CREDENTIALS] =             { "CREDENTIALS",             EXIT_STATUS_SYSTEMD },
+        [EXIT_BPF] =                     { "BPF",                     EXIT_STATUS_SYSTEMD },
```
I am wondering about using EXIT_BPF_LSM to distinguish with other future errors

iaguis replied:
I wonder that too, but since there's a limited amount of exit status available (there's 10 left) I figured one for BPF is better than a specific one...

ssahani committed:
networkd: add support for prefix allow-list and route allow-list

poettering committed:
man: extend time-{set,sync}.target + systemd-timesyncd/wait-sync docs

Let's link the three man pages together more tightly and explain what
the two targets are about, emphasizing local/quick/reliable/approximate
vs remote/slow/unreliable/accurate synchronization.

Follow-up for: 1431b2f701f7ba71e5d8664e709b1fad26797918 fe934b42e480473afba8a29a4a0d3d0e789543ac

iaguis committed:
test/fuzz: add RestrictFileSystems= directive

iaguis committed:
missing: add several filesystems

They were failing in the CI.

bluca committed:
test: update test_create_image

Run mask_supporting_services, which was missing, and remove setup_nspawn_root which
is now already called outside test_create_image

bluca committed:
test: use pkg-config to get user unit dir when installing dbus user socket

Usually on Debian ROOTLIBDIR is /lib/<arch triplet>, which is not the right place.
Use pkg-config since we define it, and then fallback to /usr/lib/systemd/user which is
the canonical location.
On both Debian&friends and Fedora dbus/dbus-broker install the user socket/service
under /usr/lib/systemd/user, not /lib/systemd/systemd/user.

bluca committed:
test: check for binariers in [usr/][s]bin

The image build function greps for ExecStart lines in unit files, but some
of them (eg: systemd-firstboot) do not use a full path.
It then falls back to 'type -P' but that only works if you have the binary
installed. For optional binaries like systemd-firstboot, the installation
can then fail.

Manually check if the binary already exists in /[usr/][s]bin.

bluca committed:
test: shortcut skip if both TEST_NO_QEMU and TEST_NO_NSPAWN are set

Allows to run all tests in bulk with TEST_NO_QEMU, skipping those where it is
mandatory, without wasting time building the image.

bluca committed:
Do not run ninja in run-integration-tests.sh

It is typically ran as root, while builds are done as normal users.
It is documented anyway to build beforehand.

iaguis committed:
README: add missing BPF requirements

iaguis committed:
README: document LSM BPF requirements

jonathanunderwood committed:
cryptsetup: add support for workqueue options

This commit adds support for disabling the read and write
workqueues with the new crypttab options no-read-workqueue
and no-write-workqueue. These correspond to the cryptsetup
options --perf-no_read_workqueue and --perf-no_write_workqueue
respectively.

angdraug committed:
man/systemd-nspawn: document hashing machine name for uid base

Explicitly document the behavior introduced in #7437: when picking a new
UID shift base with "-U", a hash of the machine name will be tried
before falling back to fully random UID base candidates.

flokli committed:
network: fix typo

IPv6 privacy extensions are plural, not singular.

flokli committed:
network: fix IPv6PrivacyExtensions=kernel handling

When set to "kernel", systemd is not supposed to touch that sysctl.

5e0534f1c13cd50ec2b143a8b18156cd37e502f7, part of
https://github.com/systemd/systemd/pull/17240 forgot to handle that
case.

Fixes https://github.com/systemd/systemd/issues/18003

thom311 committed:
sd-dhcp-client: avoid "maybe-uninitialized" warning in client_timeout_resend()

When compiling with CFLAGS='-Werror=maybe-uninitialized -Og' we get a
warning about uninitialized "next_timeout" variable.

Avoid the warning by adding an (unreachable) "default" label.

Fixes: c24288d21ee9 ("sd-dhcp-client: correct dhcpv4 renew/rebind retransmit timeouts")

milovlad committed:
Update 60-sensor.hwdb

hwdb: Add accelerometer orientation quirk for the MYRIA MY8312

thaJeztah committed:
Improve instructions for debugging failing service

In situations where a service fails to start, systemd suggests the user to
use "journalctl -xe" to get details about the failure. While running this
command does provide some additional details, most of the information is
similar to what was already printed when the service fails.

often the actual reason for the failure can be found in the logs of the
service that fails to start.

This patch updates the wording to suggest using "-u" to view the service
logs instead.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

yuwata committed:
Merge pull request #18069 from flokli/ipv6-privacy-extensions-kernel

network: fix IPv6PrivacyExtensions=kernel

yuwata committed:
network: drop redundant TAKE_PTR()

Follow-up for 16c89e649d248b0fc7c1dc7ad9f3323be4eac0f4.

yuwata committed:
sd-ndisc: fix indentation

yuwata committed:
network: fix condition for checking the provided gateway is assigned to link

Fix bug introduced by 221019166f315252304b3459902ead613b905de5.

yuwata committed:
network: make RouteDenyList= filter route prefix rather than gateway address

DenyList= filters provided prefixes, not router address.
So, RouteDenyLisy= should so for consistency.

Fixes 16c89e649d248b0fc7c1dc7ad9f3323be4eac0f4.

yuwata committed:
test-network: add tests for [IPv6AcceptRA] PrefixDenyList= or friends

yuwata committed:
network: rename DenyList= -> PrefixDenyList=

yuwata committed:
network: introduce RouterAllowList= and RouterDenyList= in [IPv6AcceptRA]

yuwata committed:
Merge pull request #18021 from ssahani/route-allow-list

networkd: add support for prefix allow-list and route allow-list

bluca committed:
tests: add TEST_QEMU_ONLY variable to run only tests where qemu is mandatory

Allows to split the test run in two parts. Most tests can run under
nspawn which is much faster, and they can be ran in one chunk with
TEST_NO_QEMU=1. The qemu-only tests, which are just a handful, can
be ran in another chunk with TEST_QEMU_ONLY=1.

Allows autopkgtest to be split in two parts.

bluca committed:
shared/dns: fix dlopen_idn return code check

Fixes https://github.com/systemd/systemd/issues/18078

poettering committed:
man: apply @Minoru's suggestions from code review

Co-authored-by: Alexander Batischev <eual.jp@gmail.com>

poettering committed:
man: use 'weak' and 'strong' for explaining difference between Wants= + Requires=

A minor tweak, that hopefully makes things a bit clearer, given that we
previously used "requirement dependency" when referring to Wants=, which
might be confusing given that we have Requires=

bluca committed:
Merge pull request #18048 from poettering/timesync-man-more

man: extend time-{set,sync}.target + systemd-timesyncd/wait-sync docs

bluca committed:
cryptsetup: use crypt_token_max if available

New API added upstream:

https://gitlab.com/cryptsetup/cryptsetup/-/commit/8a12f6dc2c75f8fd0c4969fbdc421895eb418072

yuwata committed:
fuzzers: set maximum length for several fuzzers

Hopefully fixes oss-fuzz#19081, oss-fuzz#21731, oss-fuzz#21755.
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=19081
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=21731
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=21755

yuwata committed:
resolve: slightly shorten dns_resource_key_compare_func()

yuwata committed:
resolve: do not compare key twice

yuwata committed:
resolve: slightly optimize dns_answer_add()

Previously, dns_answer_add() was O(n^2).
With this change dns_packet_extract() becomes ~15 times faster for some
extremal case.

Before:
```
$ time ./fuzz-dns-packet ~/downloads/clusterfuzz-testcase-minimized-fuzz-dns-packet-5631106733047808
/home/watanabe/downloads/clusterfuzz-testcase-minimized-fuzz-dns-packet-5631106733047808... ok

real    0m15.453s
user    0m15.430s
sys     0m0.007s
```

After:
```
$ time ./fuzz-dns-packet ~/downloads/clusterfuzz-testcase-minimized-fuzz-dns-packet-5631106733047808
/home/watanabe/downloads/clusterfuzz-testcase-minimized-fuzz-dns-packet-5631106733047808... ok

real    0m0.831s
user    0m0.824s
sys     0m0.006s
```

Hopefully fixes oss-fuzz#19227.
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=19227

yuwata committed:
Merge pull request #17477 from yuwata/network-drop-serialization-routing-policy-rule

network: do not serialize/deserialize routing policy rules

yuwata committed:
resolve: fix typo

yuwata committed:
siphash: introduce siphash24_compress_safe()

yuwata committed:
resolve: field size in dns resource record may be zero

yuwata committed:
network: constify several arguments

yuwata committed:
network: fix possible memory leak

When set_put() returns 0, then already stored rule will be unref()ed
from Manager.

yuwata committed:
network: always re-configure rules even if already exist

routing_policy_rule_get() in link_set_routing_policy_rules() does not
work when [RoutingPolicyRule] section does not have From= or To=.

yuwata committed:
network: set RoutingPolicyRule::family based on Family= setting

yuwata committed:
network: drop fib rules configured with Family=both

yuwata committed:
network: drop unnecessary checks

By the previous commit, the checks are not necessary any more.

yuwata committed:
test-network: add tests for fib rules with Family=both vs networkctl reload or reconfigure

endreszabo committed:
A typo fix in bootctl.c (#18106)

yuwata committed:
Merge pull request #18105 from yuwata/fuzz-fix-timeout

Fuzz fix timeout

bluca committed:
Merge pull request #18109 from yuwata/network-routing-policy-rule-family-both-fixes

network: fixes related to Family=both in [RoutingPolicyRule]

gportay committed:
docs: fix the link to boot loader specification

The boot loader specification link points to the boot loader interface
documentation.

This fixes the link to point to BOOT_LOADER_SPECIFICATION instead of
BOOTLOADER_INTERFACE which is itself.

bluca committed:
tests: add TEST_PREFER_NSPAWN variable to run as many as possible under nspawn

By default the test suite prefers qemu, and uses nspawn only if
a test specifically says it doesn't support qemu.
Add a variable to allow flipping the default, and run as many
tests under nspawn as possible.

daandemeyer committed:
CI: Update to mkosi github action to v9

lucaswerkmeister committed:
man: systemd-sysusers does not create home dir

Document that systemd-sysusers doesn’t create the specified home
directory, and point to systemd-tmpfiles as the service that can create
the home directory instead. (systemd-tmpfiles-setup.service is ordered
After=systemd-sysusers.service, so by the time systemd-tmpfiles would
create the home directory, sysusers would have created the owning user
already, and it should all work out.)

daandemeyer committed:
Move shared mkosi settings to a single file in mkosi.default.d/

Commit:
fix typo

ssahani committed:
network: Allow to set device's receive queues and transmit queues

keszybz committed:
Merge pull request #18088 from bluca/test_force_nspawn

tests: add TEST_QEMU_ONLY variable to run only tests where qemu is mandatory, TEST_PREFER_NSPAWN to run under nspawn unless unsupported, fix some testsuite issues on Debian

iaguis commented:
There's an upstream PR now. Closing.
